### PR TITLE
Cleanup for PermissionUtil and fixed user_limit check.

### PR DESCRIPTION
**Pull request by MinnDevelopment**
Changed some structure and design for the PermissionOverrideImpl internals

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java [hunk 1]
```diff
@@ -420,79 +481,165 @@ public static long getEffectivePermission(Channel channel, Role role)
...
+     *
+     * @throws IllegalArgumentException
+     *         If any of the arguments is {@code null}
+     *         or the specified entities are not from the same {@link net.dv8tion.jda.core.entities.Guild Guild}
+     *
+     * @return Primitive (unsigned) long value with the implicit permissions of the specified member in the specified channel
+     *
+     * @since  3.1
+     */
+    public static long getImplicitPermission(Channel channel, Member member)
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]: Is this code not identical to `getEffectivePermission(Channel, Member)`, minus the if statements that give the set of permissions based on `ADMINISTRATOR`, `MANAGE_CHANNEL` and `MANAGE_PERMISSIONS` ?

I ask, because if it is, shouldn't `getEffectivePermission` call `getImplicitPermission` and then just have those 2 if statements modify the returned value? Looking to remove duplicate code if possible.

File: src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java [hunk 2]
```diff
@@ -420,79 +481,165 @@ public static long getEffectivePermission(Channel channel, Role role)
...
+     *
+     * @throws IllegalArgumentException
+     *         If any of the arguments is {@code null}
+     *         or the specified entities are not from the same {@link net.dv8tion.jda.core.entities.Guild Guild}
+     *
+     * @return Primitive (unsigned) long value with the implicit permissions of the specified role in the specified channel
+     *
+     * @since  3.1
+     */
+    public static long getImplicitPermission(Channel channel, Role role)
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]: Doesn't look like this method takes into account the actual permissions granted to the Public role nor the provided role, it only checks the override information. Remember, overrides can have "inherit" which neither allows nor denies.

File: src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java [hunk 3]
```diff
@@ -420,79 +481,165 @@ public static long getEffectivePermission(Channel channel, Role role)
...
+    private static boolean isApplied(long permissions, Permission permission)
+    {
+        return (permissions & permission.getRawValue()) > 0;
+    }
+
     private static long apply(long permission, long allow, long deny)
     {
-        permission = permission | allow;    //Allow all the things that the cascade of roles allowed
-        permission = permission & (~deny);  //Deny everything that the cascade of roles denied.
+        deny &= ~allow;
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]: Why was this line added?
Just because something isn't specifically marked as allow **does not** mean it is denied.
Overrides have concept of inherit.

File: src/main/java/net/dv8tion/jda/core/handle/ChannelUpdateHandler.java [hunk 4]
```diff
@@ -115,15 +117,28 @@ protected Long handleInternally(JSONObject content)
...
+                        .filter(Objects::nonNull)
+                        .filter(user -> !containedMembers.contains(user))
+                        .forEach(member -> {
+                            changedMembers.add(member);
+                            toRemove.add(member.getUser().getIdLong());
+                        });
+
+                toRemove.forEach((id) -> {
+                    channel.getOverrideMap().remove(id);
+                    return true;
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/core/handle/ChannelUpdateHandler.java]: This is part of trove collections, when true is returned the next iteration step starts otherwise the forEach is interrupted.

Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/handle/ChannelUpdateHandler.java]: Interesting. Alrighty.

File: src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java (hunk 2 above)
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]: I'm not sure what you mean. It does take the public role into account:
```java
long permission = 0;
PermissionOverride override = channel.getPermissionOverride(role.getGuild().getPublicRole());
if (override != null)
    permission = apply(permission, override.getAllowedRaw(), override.getDeniedRaw());
if (role.equals(role.getGuild().getPublicRole()))
    return permission;
```

File: src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java (hunk 3 above)
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]: This makes sure that no bits in `deny` are also 1 in `allow`
Example:
`1010 & ~1000` is `1010 & 0111` -> `0010`
This is done because permissions that are allowed override the ones that are denied.

File: src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java (hunk 1 above)
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]: The reason I did not do this is because in `getEffectivePermission` it is able to return without knowing all implicit permissions.
e.g. it can return if the public role has `Permission.ADMINISTRATOR` but `getImplicitPermission` retrieves **all** permissions and then returns which would make `getEffectivePermission` less efficient.

File: src/main/java/net/dv8tion/jda/core/handle/ChannelUpdateHandler.java [hunk 5]
```diff
@@ -198,15 +213,28 @@ protected Long handleInternally(JSONObject content)
...
                 //Loop through all of the json defined overrides. If we find a match, remove the User or Role from our lists.
                 //Any entries remaining in these lists after this for loop is over will be removed from the Channel's overrides.
-                List<Role> collect = channel.getRoleOverrideMap().keySet().stream().filter(role -> !containedRoles.contains(role)).collect(Collectors.toList());
-                collect.forEach(role -> {
+                TLongList toRemove = new TLongLinkedList();
+                TDecorators.wrap(channel.getOverrideMap().keySet()).stream()
+                    .map(id -> channel.getGuild().getRoleById(id))
+                    .filter(Objects::nonNull)
+                    .filter(role -> !containedRoles.contains(role))
+                .forEach(role -> {
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/core/handle/ChannelUpdateHandler.java]: This was due to bad copy-paste

File: src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java (hunk 1 above)
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]: Then you should break the code up into helper methods. We needn't have so much duplicate code

File: src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java (hunk 2 above)
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]: That is only dealing with the override information. I don't see you ever taking into consideration the `role.getPermissionRaw`.

File: src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java (hunk 3 above)
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]: wouldn't it be better to just reorder the apply such that
```
permission &= ~deny
permission |= allow
```
?

Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]: That probably works too.

File: src/main/java/net/dv8tion/jda/core/managers/impl/AudioManagerImpl.java [hunk 6]
```diff
@@ -103,9 +105,15 @@ public void openAudioConnection(VoiceChannel channel)
                 return;
 
             final int userLimit = channel.getUserLimit(); // userLimit is 0 if no limit is set!
-            if (!self.hasPermission(channel, Permission.MANAGE_CHANNEL) && userLimit > 0 && userLimit <= channel.getMembers().size())
-                throw new PermissionException(Permission.MANAGE_CHANNEL,
-                        "Unable to connect to VoiceChannel due to userlimit! Requires permission MANAGE_CHANNEL to bypass");
+            final long perms = PermissionUtil.getImplicitPermission(channel, self);
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/managers/impl/AudioManagerImpl.java]: Calculate the implicit perms inside of the below if statement to avoid unnecessary permission calculations.

File: src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java [hunk 7]
```diff
@@ -259,46 +255,40 @@ public static boolean checkPermission(Guild guild, Member member, Permission...
...
 
         GuildImpl guild = (GuildImpl) channel.getGuild();
         if (!guild.equals(member.getGuild()))
             throw new IllegalArgumentException("Provided channel and member are not from the same guild!");
 
-        if (guild.getOwner().equals(member) // Admin or owner? If yes: no need to iterate
-                || guild.getPublicRole().hasPermission(Permission.ADMINISTRATOR)
-                || member.getRoles().stream().anyMatch(role -> role.hasPermission(Permission.ADMINISTRATOR)))
-            return true;
+//        if (guild.getOwner().equals(member) // Admin or owner? If yes: no need to iterate
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]: Above code should use `checkGuild`

File: src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java [hunk 8]
```diff
@@ -420,79 +432,179 @@ public static long getEffectivePermission(Channel channel, Role role)
...
-        long permission = getEffectivePermission(guild, member);
+        Args.notNull(channel, "Channel");
+        Args.notNull(role, "Role");
 
-        //override with channel-specific overrides of @everyone
-        PermissionOverride override = roleOverrides.get(guild.getPublicRole());
+        final Guild guild = role.getGuild();
+        checkGuild(channel.getGuild(), guild, "Role");
+
+        long permission = role.getPermissionsRaw() & guild.getPublicRole().getPermissionsRaw();
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]: Shouldn't this be an `|` ?

File: src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java [hunk 9]
```diff
@@ -420,79 +432,179 @@ public static long getEffectivePermission(Channel channel, Role role)
...
+        PermissionOverride override = channel.getPermissionOverride(guild.getPublicRole());
         if (override != null)
-        {
             permission = apply(permission, override.getAllowedRaw(), override.getDeniedRaw());
+        if (role.equals(guild.getPublicRole()))
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]: `role.isPublicRole()`

File: src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java [hunk 10]
```diff
@@ -420,79 +432,179 @@ public static long getEffectivePermission(Channel channel, Role role)
...
-                }
-                else
-                {
-                    allow = po.getAllowedRaw() | allow;     //Give all the stuff allowed by this Role's allow
-
-                    deny = (po.getDeniedRaw() | deny) & (~allow);  //Deny everything that this role denies.
-                    // This also rewrites the previous role's denies if this role allowed those permissions.
-                }
+                deny.accumulateAndGet(override.getDeniedRaw(), PermissionUtil::accumulate);
+                allow.accumulateAndGet(override.getAllowedRaw(), PermissionUtil::accumulate);
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]: Honestly these should probably just be in-line BiConsumers

Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]: Can you elaborate?

File: src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java (hunk 8 above)
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]: Good catch

File: src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java (hunk 10 above)
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]:
```java
getImplicitOverrides(channel, member, (a, d) ->
{
    allow.accumulateAndGet(a, PermissionUtil::accumulate);
    deny.accumulateAndGet(d, PermissionUtil::accumulate);
});
```
Do you mean something close to this?

Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java]: no, I meant the `PermissionUtil::accumulate`

`allow.accumulateAndGet((val1, val2) -> val1 | val2);`
